packages/babel-plugin-proposal-optional-chaining/src/index.js

```diff
@@ -22,6 +22,30 @@ export default declare((api, options) => {
     );
   }
 
+  /**
+   * Test if a given optional chain `path` needs to be memoized
+   * @param {NodePath} path
+   * @returns {boolean}
+   */
+  function needsMemoize(path) {
+    let optionalPath = path;
+    const { scope } = path;
+    while (
+      optionalPath.isOptionalMemberExpression() ||
+      optionalPath.isOptionalCallExpression()
+    ) {
+      const { node } = optionalPath;
+      const childPath = skipTransparentExprWrappers(
+        optionalPath.get("object") ?? optionalPath.get("callee"),
+      );
+      if (node.optional) {
+        return !scope.isStatic(childPath.node);
```

Comment:
I think we should only be returning if it's non-static. If it is static, we should continue the loop. Eg, `foo?.bar?.baz().qux` needs it.

Reply:
Good question. I have added `foo?.bar?.baz().qux` as an example. I overlooked the fact that `NodePath#get` always returns non-nullish.

If `childPath.node` is static, it must be `this`, `super` or `identifier` and we could stop here since `optionalPath.isOptionalChain()` will be no longer `true` in the next loop. The idea here is to find the last chain before `?` an optional chain. So in your example `childPath.node` will be `foo?.bar` for `foo?.bar?.baz().qux`. `foo?.bar` is non-static so the whole chain will then be wrapped in IIFE.

We could further optimize and allow `SimpleMemberExpression` for loose mode.